src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -1672,39 +1672,51 @@ public boolean isCastable(Type t, Type s, Warner warn) {
     // where
         class DisjointChecker {
             Set<Pair<ClassSymbol, ClassSymbol>> pairsSeen = new HashSet<>();
+            /* there are three cases for ts and ss:
+             *   - one is a class and the other one is an interface (case I)
+             *   - both are classes                                 (case II)
+             *   - both are interfaces                              (case III)
+             * all those cases are covered in JLS 23, section: "5.1.6.1 Allowed Narrowing Reference Conversion"
+             */
             private boolean areDisjoint(ClassSymbol ts, ClassSymbol ss) {
                 Pair<ClassSymbol, ClassSymbol> newPair = new Pair<>(ts, ss);
                 /* if we are seeing the same pair again then there is an issue with the sealed hierarchy
                  * bail out, a detailed error will be reported downstream
                  */
                 if (!pairsSeen.add(newPair))
                     return false;
-                if (isSubtype(erasure(ts.type), erasure(ss.type))) {
-                    return false;
-                }
-                // if both are classes or both are interfaces, shortcut
-                if (ts.isInterface() == ss.isInterface() && isSubtype(erasure(ss.type), erasure(ts.type))) {
-                    return false;
-                }
-                if (ts.isInterface() && !ss.isInterface()) {
-                    /* so ts is interface but ss is a class
-                     * an interface is disjoint from a class if the class is disjoint form the interface
-                     */
-                    return areDisjoint(ss, ts);
-                }
-                // a final class that is not subtype of ss is disjoint
-                if (!ts.isInterface() && ts.isFinal()) {
-                    return true;
-                }
-                // if at least one is sealed
-                if (ts.isSealed() || ss.isSealed()) {
-                    // permitted subtypes have to be disjoint with the other symbol
-                    ClassSymbol sealedOne = ts.isSealed() ? ts : ss;
-                    ClassSymbol other = sealedOne == ts ? ss : ts;
-                    return sealedOne.getPermittedSubclasses().stream().allMatch(type -> areDisjoint((ClassSymbol)type.tsym, other));
+
+                if (ts.isInterface() != ss.isInterface()) { // case I: one is a class and the other one is an interface
```

Comment:
Only for consideration: the specification does not have 3 cases, but 4, when the second asks the "areDisjoint" question with swapped arguments. The code here is equivalent to that, but might be easier to just re-do what the spec is doing, having handle the case where `ts` is a class, `ss` is an interface, and then have something along the lines:
```
} else if (ts.isInterface() && !ss.isInterface()) {
     return areDisjoint(ss, ts);
}
```

Reply:
you are right but given that these two cases can be folded into one, I mean when one is an interface an the other is a class, I think I prefer the current code. Both cases are symmetrical so I think they could naturally be folded into one